### PR TITLE
LPS-113543 Move and improve `Liferay.Util.sub` to frontend-js-web

### PR DESCRIPTION
**Pull request by kresimir-coko**
Fixes https://issues.liferay.com/browse/LPS-113543

I took the liberty of simplifying the utility, making it less defensive. 

One note: I couldn't manage to write a working test using `Liferay.Language.get()` which is the most common usage of this utility. By working, I mean that `Liferay.Language.get()` just wouldn't do its job.

I've tested this by checking if the icons are still being properly rendered/fetched, which is another common usage of this utility.

### CODE REVIEW

**Comment by kresimir-coko** [comment]
ci:test:relevant

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/sub.js [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+		).toBe(
+			'<a href="" target="_blank">Learn how</a> to tailor categories to your needs'
+		);
+	});
+
+	it('throws an error if the first argument is not a string', () => {
+		expect(() => sub(1)).toThrow(
+			new TypeError('the first argument must be a string')
+		);
+	});
```
Comment (diegonvs) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/sub.js:48]: So nice you're adding tests for this!

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/liferay.d.ts [hunk 2]
```diff
@@ -165,7 +165,7 @@ declare module Liferay {
 			data: Object
 		): void;
 
-		export function sub(langKey: string, ...args: any[]): string;
+		export function sub(string: string, ...args: any[]): string;
```
Comment (diegonvs) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/liferay.d.ts:168]: I think instead of `any[]`, it can be 

```typescript
type TSubArg = string | function
export function sub(string: string, ...args: TSubArg[]): string;
```

Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/liferay.d.ts:168]: > `type TSubArg = string | function`

Why do you think it can be a string or function? We are checking against an object and function in the util. If anything, replacing `any[]` could be `string[]`.

Comment (diegonvs) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/liferay.d.ts:168]: > Why do you think it can be a string or function?

Nice question. I don't remember, I spent 20min thinking but I don't remember. If I remember, I'll share with you this delirium 😂 😂 

you can use `string[]` instead of `any[]`

**Comment by kresimir-coko** [comment]
ci:test:sf

**Comment by kresimir-coko** [comment]
ci:test:relative

**Comment by kresimir-coko** [comment]
ci:test:sf

**Comment by kresimir-coko** [comment]
ci:test:relevant

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js [hunk 3]
```diff
@@ -0,0 +1,25 @@
...
+	if (typeof string !== 'string') {
+		throw new TypeError('the first argument must be a string');
+	}
+
+	const REGEX_SUB = /\{\s*([^|}]+?)\s*(?:\|([^}]*))?\s*\}/g;
+
+	return string.replace(REGEX_SUB, (match, key) =>
+		args[key] === undefined ? match : args[key]
+	);
+}
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js:25]: I think we should keep the original function of sub, by keeping the second arg as an array. The reason for this is because by changing this we are essentially causing a breaking change. Liferay.Lang.sub is a public API and so i am sure customers are using it and places outside of liferay-portal.

File: modules/apps/frontend-js/frontend-js-web/test/liferay/util/sub.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+import sub from '../../../src/main/resources/META-INF/resources/liferay/util/sub';
+
+describe('sub', () => {
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/test/liferay/util/sub.js:17]: Love to see tests! Great work

**Review by bryceosterhaus — CHANGES_REQUESTED** [review]
> I took the liberty of simplifying the utility, making it less defensive.

Hey I mentioned it below, but we should keep the exact same logic as before so that we don't introduce any breaking changes. All we really need to do here is copy the logic and then we can also convert it to TS

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js (hunk 3 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js:25]: Hmm, makes sense, that whole setup was confusing/weird to me and might have even been bugged. Doesn't this only return one argument: `data = Array.prototype.slice.call(arguments, 1);`? 
```js
const arguments = ['foo', 'bar', 'baz']

sub('foo-x', arguments); // returns 'foo-foo'
```
This is my problem with switching contexts, but IIRC, that's why I removed that whole array handling. But I can keep it in, for safety.

Comment (bryceosterhaus) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js:25]: I believe when language keys get to the client, they transform from 'foo-x' to 'Foo {0}'.

Check out the original code working here, https://codesandbox.io/s/beautiful-elion-nli1ng?file=/src/index.js

Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js:25]: Then I remembered wrong, a lot of these deprecations going on at the same time 🤷 

Thanks for catching it <3

Comment (bryceosterhaus) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js:25]: And just FYI, most of the issues in this epic are aimed at removing AUI and YUI from the util, rather than deprecating the utility all together. I just noticed that in your other PR.  Let me know if you have any questions about it

Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js:25]: So far, some of them had no AUI/YUI in them, like this `sub` one. Shouldn't we still move them away from `aui-js-web`?

**Comment by kresimir-coko** [comment]
Hey @bryceosterhaus I've brought back the old implementation, added a test when an array is passed in and updated the types in the definition file. Just having some doubts about keeping the  `data` part and not using `...args`, kinda doesn't seem intuitive from reading the function declaration that it can accept an N amount of arguments without the `...args` part.

**Comment by kresimir-coko** [comment]
ci:test:relevant

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js (hunk 3 above)
Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js]: Also, responding here to your [comment](https://github.com/liferay-frontend/liferay-portal/pull/1992#discussion_r832085259), to keep it in this one thread:
> but I don't think we actually want to deprecate this utility as it is a public API for developers to consume. Unless you saw somewhere specifically about deprecating this.

Previously, when we did these deprecations/modernizations the goal was to deprecate them so that nothing breaks.

I think that our ultimate goal is to reduce the amount of utils, but we can't straight up remove them, so we deprecate them, then move the entire module to `archived` and call it a day (after a few years ofc). 

If a customer is using a util, nothing happens to them, everything stays the same as far as they're concerned, but we have moved/deprecated/modernized them, so that our life is easier.

Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js]: @bryceosterhaus pinging for above

Comment (bryceosterhaus) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js]: > @kresimir-coko
> if a utility isn't used anywhttps://github.com/liferay-frontend/liferay-portal/pull/1992 or in one to two places, deprecate it and rewrite its usage close to source (doing this here)

So we can't actually rely just on where it is used in `liferay-portal` because they are public APIs for customers to use. So we should expect that these utils are used frequently in customer code as well, which means we shouldn't just deprecate based on how many times it is used in liferay-portal. This also goes back to the original purpose of the epic, which is not about deprecating utilities, it is focused on removing AUI.

> @kresimir-coko 
> I think that our ultimate goal is to reduce the amount of utils

I don't think this is actually the goal of the epic. The goal of the epic is to remove AUI.
From the epic, the goal is stated:

> LPS-113522
> to remove all dependency on AUI so that we can turn off frontend-js-aui-web by default, and for any functionality that needs to live on, we want it ported into a "modernized" format (ie. remove dependency on legacy APIs and libraries, pay off technical debt etc).

And from the this task in particular, the goal is stated

> LPS-113543
> Purpose of this task is to contribute to our goal of removing our dependency on AUI/YUI from liferay portal

Deprecating a util means we intend to entirely remove it in the future, which for most of these utils, we don't actually plan on removing them at all, rather we just want to remove the dependency on AUI.


So in conclusion, these tasks should be focused on AUI removal and moving them out of aui-web, if necessary. Some of these utils may not even need changes, but we at least need to verify if they do.

Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js]: Well put @bryceosterhaus, makes sense, I'll switch up my approach 👍

Comment (bryceosterhaus) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js]: Sounds good! I am happy to talk through it more too if needed, we can always jump on a call as well.

Comment (kresimir-coko) [reply modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js]: I'll see with the next couple tasks and if I have more questions then we can, for sure 😄

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by john-co** [comment]
ci:stop

**Comment by john-co** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
@magjed4289 can you give this failure a quick glance please? I think it's something unrelated. Don't go too deep, if there's a chance it's related let me know and I'll dig deeper.

**Comment by kresimir-coko** [comment]
ci:test:relevant

File: modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/custom/form/components/PartialResults.tsx [hunk 5]
```diff
@@ -73,11 +73,11 @@ const PartialResults: React.FC<IProps> = ({reportDataURL}) => {
 								{totalItems === 1
 									? Liferay.Util.sub(
 											Liferay.Language.get('x-entry'),
-											[totalItems]
+											totalItems
```
Comment (bryceosterhaus) [inline modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/custom/form/components/PartialResults.tsx:76]: These changes don't look right, as we don't want to change how the `sub(` API works

File: modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/custom/form/components/PartialResults.tsx [hunk 6]
```diff
@@ -73,11 +73,11 @@ const PartialResults: React.FC<IProps> = ({reportDataURL}) => {
...
 									  )
 									: Liferay.Util.sub(
 											Liferay.Language.get('x-entries'),
-											[totalItems]
+											totalItems
```
Comment (bryceosterhaus) [inline modules/apps/data-engine/data-engine-js-components-web/src/main/resources/META-INF/resources/js/custom/form/components/PartialResults.tsx:80]: ^ same as above

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+export default function sub(string, data) {
+	const REGEX_SUB = /\{\s*([^|}]+?)\s*(?:\|([^}]*))?\s*\}/g;
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js:16]: This is a constant, so it can be moved up and out of the function. Otherwise, if we leave it here we have to create this regular expression on every all to `sub`

File: modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+	if (typeof string !== 'string') {
+		throw new TypeError('the first argument must be a string');
+	}
```
Comment (bryceosterhaus) [inline modules/apps/frontend-js/frontend-js-web/src/main/resources/META-INF/resources/liferay/util/sub.js:20]: This feels a bit redundant to our types file. I don't think it is necessary to add runtime type checking here, especially because it didn't exist here initially.

**Review by bryceosterhaus — CHANGES_REQUESTED** [review]
Just had a new things to note here.

One additional thing, have we considered moving some of these utils to TS files? I wasn't sure if they needed to be js files due to the global.es.js import.

**Comment by kresimir-coko** [comment]
> have we considered moving some of these utils to TS files?

It has crossed my mind, but I shrugged it off, we do add types to the global type definition, but that's not the same thing. We can start doing it halfway through the epic if you think it's a good idea.

**Comment by kresimir-coko** [comment]
@bryceosterhaus I've squashed old commits and added new ones 🙇

**Comment by bryceosterhaus** [comment]
@kresimir-coko for the TS stuff, I wouldn't worry about it yet. Adding them to the definitions file for now should be fine.

**Review by bryceosterhaus — COMMENTED** [review]
LGTM

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by diegonvs** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Comment by kresimir-coko** [comment]
ci:test:relevant

**Review by bryceosterhaus — APPROVED** [review]
ci:forward

**Comment by kresimir-coko** [comment]
ci:forward

**Comment by kresimir-coko** [comment]
ci:forward
